### PR TITLE
Create ACL bindings for kafka

### PR DESCRIPTION
**Pull request by jackdelahunt**
This pr adds the functionality of creating ACL bindings for any kafka instance, it adds more options within the kafka resource to achieve this. It also adds a acl resource to use with the need for the kafka resource to be created at the same time.

### Terraform config (main.tf)
```
terraform {
  required_providers {
    rhoas = {
      source  = "registry.terraform.io/redhat-developer/rhoas"
      version = "0.1.0"
    }
  }
}

provider "rhoas" {
    offline_token = "..."
}
resource "rhoas_service_account" "srvcaccnt" {
  name = "service_account"
}

resource "rhoas_kafka" "instance" {
  name = "instance"
  acl = [
    {
      principal = rhoas_service_account.srvcaccnt.client_id,
      resource_type = "TOPIC",
      resource_name = "topic-1",
      pattern_type = "LITERAL",
      operation_type = "ALL",
      permission_type = "ALLOW",
    },
  ]
}

resource "rhoas_topic" "topic-1" {
  kafka_id = rhoas_kafka.instance.id
  name = "topic-1"
  partitions = 1
}

resource "rhoas_topic" "topic-2" {
  kafka_id = rhoas_kafka.instance.id
  name = "topic-2"
  partitions = 1
}

resource "rhoas_acl" "acl" {
  kafka_id = rhoas_kafka.instance.id
  principal = rhoas_service_account.srvcaccnt.client_id
  resource_type = "TOPIC"
  resource_name = "topic-2"
  pattern_type = "LITERAL"
  operation_type = "ALL"
  permission_type = "ALLOW"
}
```
### Commands
- `make install`
- `terraform init`
- `terraform apply`

### Expected Results
- New kafka created named "instance"
- Kafka has 2 topics named "topic-1" and "topic-2"
- New service account created called "service_account" (description in UI)
- Service account has access to all actions for these to topics in the kafka

### CODE REVIEW

File: rhoas/kafkas/resource_kafka.go [hunk 1]
```diff
@@ -89,6 +90,15 @@ func ResourceKafka() *schema.Resource {
 				Type:        schema.TypeString,
 				Computed:    true,
 			},
+			"acl": {
+				Type:     schema.TypeList,
+				ForceNew: true,
+				Required: true,
```
Comment (dimakis) [inline rhoas/kafkas/resource_kafka.go]: Does this mean that an ACL is required when provisioning a new kafka instance?

File: rhoas/kafkas/resource_kafka.go [hunk 2]
```diff
@@ -249,9 +254,85 @@ func kafkaCreate(ctx context.Context, d *schema.ResourceData, m interface{}) dia
...
+		}
+
+		principal, ok := element["principal"].(string)
+		if !ok {
+			return errors.Errorf("There was a problem getting the principal value in the kafka acl")
+		}
+
+		// required for api, the user id, service account id or * works
+		// when appended to User:
+		principal = "User:" + principal
```
Comment (dimakis) [inline rhoas/kafkas/resource_kafka.go]: is there any validation to check if the user passes the principal in as `User:blah` or `user:blah` or `user: blah`, you get my point, in the TF script?
in those cases wouldn't the principal come out as `User:User:blah` ?

**Review by dimakis — COMMENTED** [review]
Looks good, I just have a couple of comments / questions

File: rhoas/kafkas/resource_kafka.go (hunk 1 above)
Comment (jackdelahunt) [reply rhoas/kafkas/resource_kafka.go]: It is required that the acl is passed but no acls needs to be in the list. For example this is a valid kafka instance resource.

```
resource "rhoas_kafka" "instance" {
    name = "instance"
    acl = []
}
```

File: rhoas/kafkas/resource_kafka.go (hunk 2 above)
Comment (jackdelahunt) [reply rhoas/kafkas/resource_kafka.go]: I think we just need to set a standard for these inputs and consider anything outside of that not valid. If we define that the user must enter the client_id or user id in the principal field then starting it with User: is automatically not supported.

File: rhoas/kafkas/resource_kafka.go (hunk 1 above)
Comment (dimakis) [reply rhoas/kafkas/resource_kafka.go]: ah very good, so if the acl is not passed then a list of `blank` acls is created.


so in the case where we want to create a kafka, with no acls just yet, we'ld have to pass in the empty list?
could a check be made in the code to see if the acl list is passed in? if not create the empty acl list in the code as opposed to having the user doing it?

File: rhoas/kafkas/resource_kafka.go (hunk 2 above)
Comment (dimakis) [reply rhoas/kafkas/resource_kafka.go]: my preference here would be to have a little validation around the prepending of `User:` here but that is just my own opinion

Comment (jackdelahunt) [reply rhoas/kafkas/resource_kafka.go]: I would normally agree but... then doing what I am in this pr not impossible but a mess, I am using a service account client id as the principle but I would need to prepend it with User:, this afaik is possible but a pain in terraform

File: rhoas/kafkas/resource_kafka.go [hunk 3]
```diff
@@ -249,9 +254,85 @@ func kafkaCreate(ctx context.Context, d *schema.ResourceData, m interface{}) dia
...
+
+		resourceType, ok := element["resource_type"].(string)
```
Comment (dimakis) [inline rhoas/kafkas/resource_kafka.go]: This value needs to be all caps (uppercase) for a call to work correctly right? could you use `ToUpper` func here to allow for a better experience?

The same can be said for `pattern`, `operation` and `permission`

Comment (jackdelahunt) [reply rhoas/kafkas/resource_kafka.go]: done 👍

File: rhoas/kafkas/resource_kafka.go (hunk 1 above)
Comment (jackdelahunt) [reply rhoas/kafkas/resource_kafka.go]: done 👍

File: rhoas/kafkas/resource_kafka.go (hunk 2 above)
Comment (jackdelahunt) [reply rhoas/kafkas/resource_kafka.go]: now i have realized doing this changing of values can cause issues with terraform syncing after create, i avoided it in this pr in a hacky way but something to think about for the future.

File: rhoas/kafkas/resource_kafka.go [hunk 4]
```diff
@@ -265,7 +265,13 @@ func kafkaCreate(ctx context.Context, d *schema.ResourceData, m interface{}) dia
 
 func createACLForKafka(ctx context.Context, api rhoasAPI.Clients, d *schema.ResourceData, kafka *kafkamgmtclient.KafkaRequest) error {
 
-	acl, ok := d.Get("acl").([]interface{})
+	aclInput := d.Get("acl")
```
Comment (dimakis) [inline rhoas/kafkas/resource_kafka.go:269]: great changes

**Review by dimakis — APPROVED** [review]
LGTM, 
Great work @jackdelahunt

**Comment by wtrocki** [comment]
Looks good. Sorry for late review

**Pull request merged**